src/tests/test_blk_read_tracker.cpp

```diff
@@ -352,6 +358,198 @@ TEST_F(BlkReadTrackerTest, TestInsRmWithWaiterOverlapMultiReads2) {
+    }
+
+    LOGINFO("Step 4: threaded remove issued.");
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    LOGINFO("Step 4: all threads joined.");
+}
+
+TEST_F(BlkReadTrackerTest, TestThreadedInsertWaitonThenRemove) {
+    auto align = 8ul;
+    LOGINFO("Step 1: set entries per record to {}.", align);
+    get_inst()->set_entries_per_record(align);
+
+    std::vector< BlkId > bids{{12, 6, 0}, {18, 5, 0}, {25, 8, 0}, {36, 16, 0}, {57, 4, 0}, {66, 18, 0}, {92, 14, 0}};
+    const auto repeat = 100ul;
+    std::vector< std::thread > op_threads;
+    for (const auto& b : bids) {
+        std::thread t([this, &b]() {
+            for (auto j = 0ul; j < repeat; ++j) {
+                get_inst()->insert(b);
+            }
+        });
+        op_threads.push_back(std::move(t));
+    }
+
+    LOGINFO("Step 2: threaded insert issued.");
+
+    std::vector< bool > called(bids.size(), false);
+    std::mutex mtx;
+    for (auto i = 0ul; i < bids.size(); ++i) {
+        std::thread t([this, &bids, i, &mtx, &called]() {
+            get_inst()->wait_on(bids[i], [i, &bids, &mtx, &called]() {
+                std::unique_lock lk(mtx);
+                LOGMSG_ASSERT(called[i] == false, "not expecting callback to be called more than once!");
+                called[i] = true;
+                LOGINFO("wait_on called on blkid: {};", bids[i].to_string());
+            });
+        });
+        op_threads.push_back(std::move(t));
+    }
+
+    // wait for all insert to complete, otherwise it will race with remove thread;
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    op_threads.clear();
+
+    LOGINFO("Step 3: threaded wait_on issued on all bids.");
+
+    for (const auto& b : bids) {
+        for (auto j = 0ul; j < repeat; ++j) {
+            std::thread t([this, &b]() { get_inst()->remove(b); });
+            op_threads.push_back(std::move(t));
+        }
+    }
+
+    LOGINFO("Step 3: threaded remove issued.");
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    LOGINFO("Step 4: all threads joined.");
+
+    for (const auto x : called) {
+        LOGMSG_ASSERT(x == true, "expecting all waiters to be called");
+    }
+
+    LOGINFO("Step 5: all bids wait_on cb called.");
+}
+
+/*
+ * Purpose of this test:
+ * 1. Concurrent random insert/remove/wait_on operations in different threads.
+ *
+ * */
+TEST_F(BlkReadTrackerTest, TestThreadedInsertRemoveAndWait2) {
+    auto align = 8ul;
+    LOGINFO("Step 1: set entries per record to {}.", align);
+    get_inst()->set_entries_per_record(align);
+
+    std::mutex mtx;
+    std::list< BlkId > inserted_bids;
+
+    std::atomic< uint32_t > outstanding_wait_bids_cnt = 0ul;
+
+    LOGINFO("Step 2: randome threaded insert/remove/wait_on operation:");
+    std::list< std::thread > op_threads;
+    auto nitr = 0ul;
+    while (nitr++ < 200ul || inserted_bids.empty() == false) {
+        std::thread t([this, &outstanding_wait_bids_cnt, &nitr, &inserted_bids, &mtx]() {
```

Comment:
Is it more stress testing to create 200 threads and each doing one operation vs creating N threads and each doing M iteration of random operations

Reply:
the purpose of the test is to create 200 threads, doing more iterations in a single thread won't help create racing window.